actors/builtin/multisig/multisig_test.go

```diff
@@ -311,6 +310,359 @@ func TestApprove(t *testing.T) {
+
+			initialSigners:   []addr.Address{anne, bob, chuck},
+			initialApprovals: int64(3),
+
+			addSigner: chuck,
+			increase:  false,
+
+			expectSigners:   []addr.Address{anne, bob, chuck},
+			expectApprovals: int64(3),
+			code:            exitcode.ErrPlaceholder,
+		},
+	}
+
+	builder := mock.NewBuilder(context.Background(), t, multisigWalletAdd).WithCaller(builtin.InitActorAddr, builtin.InitActorCodeID)
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			rt := builder.Build()
+
+			actor.constructAndVerify(rt, tc.initialApprovals, noUnlockDuration, tc.initialSigners...)
+
+			rt.SetCaller(multisigWalletAdd, builtin.AccountActorCodeID)
+			rt.ExpectValidateCallerAddr(multisigWalletAdd)
+			if tc.code != exitcode.Ok {
+				rt.ExpectAbort(tc.code, func() {
+					actor.addSigner(rt, tc.addSigner, tc.increase)
+				})
+			} else {
+				actor.addSigner(rt, tc.addSigner, tc.increase)
+				var st multisig.MultiSigActorState
+				rt.Readonly(&st)
+				assert.Equal(t, tc.expectSigners, st.Signers)
+				assert.Equal(t, tc.expectApprovals, st.NumApprovalsThreshold)
+			}
+			rt.Verify()
+		})
+	}
+}
+
+type removeSignerTestCase struct {
+	desc string
+
+	initialSigners   []addr.Address
+	initialApprovals int64
+
+	removeSigner addr.Address
+	decrease     bool
+
+	expectSigners   []addr.Address
+	expectApprovals int64
+	code            exitcode.ExitCode
+}
+
+func TestRemoveSigner(t *testing.T) {
+	actor := msActorHarness{multisig.MultiSigActor{}, t}
+
+	multisigWalletAdd := newIDAddr(t, 100)
+	anne := newIDAddr(t, 101)
+	bob := newIDAddr(t, 102)
+	chuck := newIDAddr(t, 103)
+	richard := newIDAddr(t, 104)
+
+	const noUnlockDuration = int64(0)
+
+	testCases := []removeSignerTestCase{
+		{
+			desc: "happy path remove signer",
+
+			initialSigners:   []addr.Address{anne, bob, chuck},
+			initialApprovals: int64(2),
+
+			removeSigner: chuck,
+			decrease:     false,
+
+			expectSigners:   []addr.Address{anne, bob},
+			expectApprovals: int64(2),
+			code:            exitcode.Ok,
+		},
+		{
+			desc: "remove signer and decrease threshold",
+
+			initialSigners:   []addr.Address{anne, bob, chuck},
+			initialApprovals: int64(2),
+
+			removeSigner: chuck,
+			decrease:     true,
+
+			expectSigners:   []addr.Address{anne, bob},
+			expectApprovals: int64(1),
+			code:            exitcode.Ok,
+		},
+		{
+			desc: "remove signer with automatic threshold decrease",
+
+			initialSigners:   []addr.Address{anne, bob, chuck},
+			initialApprovals: int64(3),
+
+			removeSigner: chuck,
+			decrease:     false,
+
+			expectSigners:   []addr.Address{anne, bob},
+			expectApprovals: int64(2),
+			code:            exitcode.Ok,
+		},
+		{
+			desc: "remove signer from single singer list",
+
+			initialSigners:   []addr.Address{anne},
+			initialApprovals: int64(2),
+
+			removeSigner: anne,
+			decrease:     false,
+
+			expectSigners:   nil,
+			expectApprovals: int64(1),
+			code:            exitcode.Ok,
+		},
```

Comment:
Want to draw attention to this test: This doesn't seem like the right behaviour -- the Multisig actor should probably not allow the only signer to be removed.

Reply:
Please file and link an issue.